tvm/python/tvm/topi/adreno/conv2d_nhwc.py

```diff
@@ -0,0 +1,301 @@
+            axis=[ry, rx, rcc, rcb],
+        ),
+        tag="conv2d_nhwc",
+    )
+
+    dummy_cast = te.compute((batch, out_height_orig, out_width_orig, num_filter_chunk, num_filter_block), lambda n,y,x,fc,fb: conv[n,y,x,fc,fb].astype(out_dtype), tag="dummy_cast")
+    return te.compute((batch, out_height_orig, out_width_orig, out_channels), lambda n,y,x,c: dummy_cast[n,y,x,c//4,c%4], tag="cast_from_acc" + args["accumulator"][-2:])
+
+def schedule_conv2d_NHWC(cfg, s, output, args={}):
+    """schedule optimized for batch size = 1"""
+    dummy = output.op.input_tensors[0]
+    conv = dummy.op.input_tensors[0]
+    #latest = s.outputs[0].output(0)
+
+    ##### space definition begin #####
+    n, y, x, fc, fb = s[conv].op.axis
+    ry, rx, rcc, rcb = s[conv].op.reduce_axis
+    cfg.define_split("tile_fc", fc, num_outputs=3,
+                filter=lambda entity: entity.size[1] <= 16 and entity.size[2] >= 2 and entity.size[2] < 128 )
+    cfg.define_split("tile_y", y, num_outputs=3,
+                filter=lambda entity: entity.size[1] <= 16 and entity.size[2] < 16 )
+    cfg.define_split("tile_x", x, num_outputs=3,
+                filter=lambda entity: entity.size[1] <= 16 and entity.size[2] < 16 )
+
+    cfg.define_split("tile_rc", rcc, num_outputs=2)
+    cfg.define_split("tile_ry", ry, num_outputs=2)
+    cfg.define_split("tile_rx", rx, num_outputs=2)
+    cfg.define_knob("auto_unroll_max_step", [0, 512, 1500])
+    cfg.define_knob("unroll_explicit", [0, 1])
+
+    cfg.multi_filter(filter=lambda entity: entity["tile_fc"].size[2] * entity["tile_y"].size[2] * entity["tile_x"].size[2] in range(32,1024))
+    ##### space definition end #####
+
+    #if autotvm.GLOBAL_SCOPE.in_tuning:
+    #    pad_data, flattened_kernel = s[conv].op.input_tensors
+    #    kernel = s[flattened_kernel].op.input_tensors[0]
+    #    s[flattened_kernel].compute_inline()
+    #else:
+    #    pad_data, kernel = s[conv].op.input_tensors
+    #    flattened_kernel = kernel
+
+    pad_data, kernel = s[conv].op.input_tensors
+    s[kernel].compute_inline()
+    s[pad_data].compute_inline()
+    #kernel = flattened_kernel
+
+    pack_data = pad_data.op.input_tensors[0]
+    s[pack_data].compute_inline()
+
+    ### conv only
+    ##if conv.op in s.outputs:
+    ##    output = conv
+    ##    OL = s.cache_write(conv, "local")
+    ### conv -> output (e.g. when casting conv output)
+    ##elif output.op in s.outputs:
+    ##    output = s.outputs[0].output(0)
+    ##    s[conv].set_scope("local")
+    ##    OL = conv
+    ### conv -> injective -> ... -> injective -> output
+    ##else:
+    ##    # Explicitly mark the output cast to be computed inline
+    ##    # the other injective ops are inlined via traverse_inline.
+    ##   s[output].compute_inline()
+    latest = s.outputs[0].output(0)
+    #s[conv].set_scope("local")
+
+    # create cache stage
+    def get_texture_storage(shape):
+        limit = 16384
```

Comment:
Instead of magic number here. Is it possible to define some global variable with meaningful name and use here? Is this number not retrievable from querying OpenCL API?

Reply:
Done. Yes. This value is a minimal value of max width/height for 2D image specified in OpenCL spec.

See `CL_​DEVICE_​IMAGE2D_​MAX_​WIDTH` in https://www.khronos.org/registry/OpenCL/sdk/2.2/docs/man/html/clGetDeviceInfo.html